File: leetcode/python/1424.diagonal-traverse-ii.py
from typing import List


class Solution:
    def findDiagonalOrder(self, nums: List[List[int]]) -> List[int]:
        if not len(nums) or not len(nums[0]):
            return []
        m = []
        for i, row in enumerate(nums):
            for j, v in enumerate(row):
                if i + j >= len(m):
                    m.append([])
                m[i + j].append(v)
        res = [v for d in m for v in reversed(d)]
        return res

File: leetcode/cpp/1424.diagonal-traverse-ii.cpp
#include "oj_header.h"

class Solution
{
public:
    vector<int> findDiagonalOrder(vector<vector<int>> &nums)
    {
        if (nums.empty() || nums[0].empty())
            return {};

        vector<vector<int>> m;

        for (size_t row = 0; row < nums.size(); row++)
        {
            for (size_t col = 0; col < nums[row].size(); col++)
            {
                if (row + col >= m.size())
                {
                    m.push_back({});
                }
                m[row + col].push_back(nums[row][col]);
            }
        }

        vector<int> v{};
        for (auto &line : m)
        {
            v.insert(end(v), rbegin(line), rend(line));
        }

        return v;
    }
};
